### PR TITLE
Add S3 config and utils.

### PR DESCRIPTION
**Pull request by JinheLin**
### What problem does this PR solve?

Issue Number: ref #6827

### What is changed and how it works?

1. Add more config related to S3 storage.
2. Refine `S3Client` creation: we can share a S3Client to avoid creating new S3Client object every time.
### Check List

Tests 

- [ ] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
None
```

### CODE REVIEW

File: dbms/src/Server/StorageConfigParser.h [hunk 1]
```diff
@@ -99,9 +99,16 @@ struct StorageS3Config
     String bucket;
     String access_key_id;
     String secret_access_key;
+    UInt64 max_connections = 1024;
+    UInt64 connection_timeout_ms = 1000;
+    UInt64 request_timeout_ms = 3000;
+    String cache_dir;
+    UInt64 cache_capacity = 0;
+    UInt64 cache_strategy = 0;
```
Comment (JaySon-Huang) [inline dbms/src/Server/StorageConfigParser.h:107]: Add some comments about the "strategy"?

File: dbms/src/Storages/S3/S3Common.cpp [hunk 2]
```diff
@@ -126,10 +126,12 @@ void ClientFactory::init(const StorageS3Config & config_)
     config = config_;
     Aws::InitAPI(aws_options);
     Aws::Utils::Logging::InitializeAWSLogging(std::make_shared<AWSLogger>());
+    shared_client = create();
 }
 
 void ClientFactory::shutdown()
 {
+    shared_client.reset(); // Reset S3Client before Aws::ShutdownAPI.
```
Comment (JaySon-Huang) [inline dbms/src/Storages/S3/S3Common.cpp:134]: Is it a must-do that all S3Clients are released before Aws::ShutdownAPI?

As `shared_client` is a shared_ptr, I think it maybe hard for us to ensure all threads are stopped handling the S3 requests before ClientFactory::shutdown.

**Review by JaySon-Huang — APPROVED** [review]
LGTM

File: dbms/src/Storages/S3/S3Common.cpp [hunk 3]
```diff
@@ -144,13 +146,53 @@ ClientFactory & ClientFactory::instance()
...
+const String & ClientFactory::bucket() const
+{
+    // `bucket` is read-only.
+    return config.bucket;
+}
+
+std::shared_ptr<Aws::S3::S3Client> ClientFactory::sharedClient() const
+{
+    // `shared_client` is created during initialization and destroyed when process exits
+    // which means it is read-only when processing requests. So, it is saft to read `shared_client`
```
Comment (Lloyd-Pottiger) [inline dbms/src/Storages/S3/S3Common.cpp:161]:
```suggestion
    // which means it is read-only when processing requests. So, it is safe to read `shared_client`
```

**Review by Lloyd-Pottiger — APPROVED** [review]

**Pull request merged**